x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsManager.java

```diff
@@ -174,40 +126,56 @@ private void createStatsIndexAndUpdateMappingsIfNecessary(Client client, Cluster
-            task.markAsCompleted();
-            return;
-        }
-        DataFrameAnalyticsTaskState reindexingState = new DataFrameAnalyticsTaskState(DataFrameAnalyticsState.REINDEXING,
-            task.getAllocationId(), null);
+    private void determineProgressAndResume(DataFrameAnalyticsTask task, DataFrameAnalyticsConfig config) {
         DataFrameAnalyticsTask.StartingState startingState = DataFrameAnalyticsTask.determineStartingState(
             config.getId(), task.getParams().getProgressOnStart());
 
         LOGGER.debug("[{}] Starting job from state [{}]", config.getId(), startingState);
         switch (startingState) {
             case FIRST_TIME:
-                task.updatePersistentTaskState(reindexingState, ActionListener.wrap(
-                    updatedTask -> reindexDataframeAndStartAnalysis(task, config),
-                    task::setFailed
-                ));
+                executeStep(task, config, new ReindexingStep(clusterService, client, task, auditor, config));
                 break;
             case RESUMING_REINDEXING:
-                task.updatePersistentTaskState(reindexingState, ActionListener.wrap(
-                    updatedTask -> executeJobInMiddleOfReindexing(task, config),
-                    task::setFailed
-                ));
+                executeJobInMiddleOfReindexing(task, config);
                 break;
             case RESUMING_ANALYZING:
-                startAnalytics(task, config);
+                executeStep(task, config, new AnalysisStep(client, task, auditor, config, processManager));
                 break;
             case FINISHED:
             default:
                 task.setFailed(ExceptionsHelper.serverError("Unexpected starting state [" + startingState + "]"));
         }
     }
 
+    private void executeStep(DataFrameAnalyticsTask task, DataFrameAnalyticsConfig config, DataFrameAnalyticsStep step) {
+        task.setStep(step);
+
+        ActionListener<StepResponse> stepListener = ActionListener.wrap(
+            stepResponse -> {
+                if (stepResponse.isTaskComplete()) {
+                    LOGGER.info("[{}] Marking task completed", config.getId());
+                    task.markAsCompleted();
+                    return;
+                }
+                switch (step.name()) {
+                    case REINDEXING:
+                        executeStep(task, config, new AnalysisStep(client, task, auditor, config, processManager));
+                        break;
+                    case ANALYSIS:
+                        // This is the last step
+                        LOGGER.info("[{}] Marking task completed", config.getId());
+                        task.markAsCompleted();
+                        break;
+                    default:
+                        task.markAsFailed(ExceptionsHelper.serverError("Unknown step [{}]", step));
+                }
```

Comment:
it almost seems to me that the `stepResponse` should have this info, or maybe the step itself? Doing a switch case here seems to be missing helpful polymorphism

Reply:
My thinking was that each step is isolated and doesn't know about the other steps. Then the manager's responsibility is to execute the steps in the right order. So this is what is done here. Depending on which step just completed, the manager starts the next appropriate step.